### PR TITLE
refactor: enable eslint hook rules

### PR DESCRIPTION
**Pull request by frosso**
Fixes #

#### Changes proposed in this Pull Request

Discussed on Slack: p1621001462390700-slack-CGGCLBN58

Enabling the following eslint rules:
- `react-hooks/exhaustive-deps` : ensures that all the necessary dependencies are passed to hooks
- `react-hooks/rules-of-hooks` : ensures that hooks aren't called conditionally or within other hooks

Refactored the code (where I knew I could do it with confidence) to fix the rules.

#### Testing instructions

Most of the changes affect the "grouped settings" work, behind a feature flag:
- Ensure you have the "grouped settings" flag enabled in WCPay Dev Tools
- Visit http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout

-------------------

- [ ] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

### CODE REVIEW

File: client/components/page/index.js [hunk 1]
```diff
@@ -16,9 +16,13 @@ const Page = ( { children, maxWidth, isNarrow, className = '' } ) => {
 		classNames.push( 'is-narrow' );
 	}
 
-	const fraudScriptsConfig =
```
Comment (frosso) [inline client/components/page/index.js:19]: Moved within the hook - same difference

File: client/data/settings/hooks.js [hunk 2]
```diff
@@ -10,46 +10,61 @@ import { useSelect, useDispatch } from '@wordpress/data';
  */
 import { STORE_NAME } from '../constants';
 
-export const useEnabledPaymentMethodIds = () =>
-	useSelect( ( select ) => {
-		const { getEnabledPaymentMethodIds } = select( STORE_NAME );
-		const { updateEnabledPaymentMethodIds } = useDispatch( STORE_NAME );
```
Comment (frosso) [inline client/data/settings/hooks.js:16]: `useDispatch` was called within `useSelect` - see https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

File: client/data/settings/hooks.js [hunk 3]
```diff
@@ -10,46 +10,61 @@ import { useSelect, useDispatch } from '@wordpress/data';
...
-		return {
-			enabledPaymentMethodIds: getEnabledPaymentMethodIds(),
-			updateEnabledPaymentMethodIds,
-		};
-	} );
-
-export const useGeneralSettings = () =>
-	useSelect( ( select ) => {
-		const { getIsWCPayEnabled } = select( STORE_NAME );
-		const { updateIsWCPayEnabled } = useDispatch( STORE_NAME );
```
Comment (frosso) [inline client/data/settings/hooks.js:27]: Same as above

File: client/data/settings/hooks.js [hunk 4]
```diff
@@ -10,46 +10,61 @@ import { useSelect, useDispatch } from '@wordpress/data';
...
-			getSettings,
-			hasFinishedResolution,
-			isResolving,
-			isSavingSettings,
-		} = select( STORE_NAME );
-
-		const isLoading =
-			isResolving( 'getSettings' ) ||
-			! hasFinishedResolution( 'getSettings' );
-		const { saveSettings } = useDispatch( STORE_NAME );
```
Comment (frosso) [inline client/data/settings/hooks.js:47]: Same as above

File: client/deposits/list/index.js [hunk 5]
```diff
@@ -74,10 +74,8 @@ export const DepositsList = () => {
 		getQuery()
 	);
 
-	const columnsArgs = [
```
Comment (frosso) [inline client/deposits/list/index.js:77]: I know this was a neat thing to do, but the "exhaustive deps" rule cannot determine the dependencies when used this way.

File: client/payment-details/timeline/index.js [hunk 6]
```diff
@@ -15,11 +15,8 @@ import mapTimelineEvents from './map-events';
 import Loadable, { LoadableBlock } from 'components/loadable';
 
 const PaymentDetailsTimeline = ( { chargeId } ) => {
-	if ( ! wcpaySettings.featureFlags.paymentTimeline ) {
```
Comment (frosso) [inline client/payment-details/timeline/index.js:18]: Moved to `client/payment-details/index.js` because otherwise the `useTimeline` hook was called conditionally

File: client/payment-details/index.js [hunk 7]
```diff
@@ -44,7 +44,9 @@ const PaymentDetails = ( props ) => {
 		<Page maxWidth={ 1032 } className="wcpay-payment-details">
 			{ testModeNotice }
 			<PaymentDetailsSummary charge={ charge } isLoading={ isLoading } />
-			<PaymentDetailsTimeline chargeId={ chargeId } />
```
Comment (frosso) [inline client/payment-details/index.js:47]: Moved from `client/payment-details/timeline/index.js`

File: client/transactions/list/index.js [hunk 8]
```diff
@@ -149,14 +149,15 @@ export const TransactionsList = ( props ) => {
 		isLoading: isSummaryLoading,
 	} = useTransactionsSummary( getQuery(), props.depositId );
 
-	const columnsArgs = [
```
Comment (frosso) [inline client/transactions/list/index.js:152]: I know this was a neat thing to do, but the "exhaustive deps" rule cannot determine the dependencies when used this way.

File: client/utils/use-confirm-navigation.js [hunk 9]
```diff
@@ -14,11 +14,17 @@ import { getHistory } from '@woocommerce/navigation';
  * - useConfirmNavigation( saved => { if ( ! saved ) return 'Discard unsaved changes?' }, [ saved ] );
  *
  * @param {Function} getMessage returns confirmation message string if one should appear
- * @param {Array} deps effect dependencies
+ * @return {Function} The callback to execute
  */
-const useConfirmNavigation = ( getMessage, deps ) => {
```
Comment (frosso) [inline client/utils/use-confirm-navigation.js:19]: Refactored to remove `deps`, because otherwise the rule couldn't determine the dependencies to `useEffect`.
Now this function returns a callback to be used within `useEffect`. And the reference to `getMessage` is "stabilized" through `useRef`.

File: client/payment-request/blocks/use-initialization.js [hunk 10]
```diff
@@ -87,19 +95,6 @@ export const useInitialization = ( {
 		billing.cartTotalItems,
 	] );
 
-	const abortPayment = ( paymentMethod, message ) => {
```
Comment (frosso) [inline client/payment-request/blocks/use-initialization.js:90]: Moved their definition inside the `useEffect` hook - otherwise they would need to be declared as dependency and "stabilized" through `useCallback`

File: client/disputes/evidence/index.js [hunk 11]
```diff
@@ -376,19 +376,25 @@ export default ( { query } ) => {
...
 		);
-	}, [ pristine ] );
+	} );
+
+	useEffect( confirmationNavigationCallback, [
+		pristine,
+		confirmationNavigationCallback,
+	] );
 
-	const fetchDispute = async () => {
```
Comment (frosso) [inline client/disputes/evidence/index.js:381]: Moved inside the `useEffect` hook - otherwise it would have needed to be declared as dependency and stabilized with `useCallback`

File: client/disputes/evidence/index.js [hunk 12]
```diff
@@ -367,7 +367,7 @@ export default ( { query } ) => {
 			}
 		);
 
-	useConfirmNavigation( () => {
```
Comment (frosso) [inline client/disputes/evidence/index.js:370]: See `client/utils/use-confirm-navigation.js` for explanation on the slight refactor.

The effect is still called when the dependencies change.

So instead of:
```
const useCustomHook = (callback, dependencies) => {
  useEffect(() => {callback()}, dependencies);
}
...
useCustomHook(() => thing(), [dependency]);
```

Now we have:
```
const useCustomHook = (callback) => callback;
...
const callback = useCustomHook(() => thing());
useEffect(() => callback(), [callback, dependency]);
```

Same difference - but it makes it more explicit which dependencies need to run and it now has a stable reference on the callback.

File: client/transactions/list/index.js [hunk 13]
```diff
@@ -149,14 +149,15 @@ export const TransactionsList = ( props ) => {
...
-		! props.depositId,
-		wcpaySettings.isSubscriptionsActive,
-		! getQuery().orderby || 'date' === getQuery().orderby,
-	];
+	const sortByDate = ! getQuery().orderby || 'date' === getQuery().orderby;
 	const columnsToDisplay = useMemo(
-		() => getColumns( ...columnsArgs ),
```
Comment (frosso) [inline client/transactions/list/index.js:158]: Feeling confident in this change, since it would have been picked up by the snapshot tests.

**Review by Ferdev — APPROVED** [review]
I've been testing the plugin and I couldn't find anything wrong. It would be good if some other person more familiar with the deposits/transactions/disputes part could give it a spin just to be sure we're not breaking anything.

Regarding the code, looks perfect, and great job spotting this issue!!

File: .eslintrc [hunk 14]
```diff
@@ -72,8 +72,8 @@
 		"@wordpress/i18n-translator-comments":     "off",
 		"@wordpress/valid-sprintf":                "off",
 		"@woocommerce/dependency-group":           "off",
-		"react-hooks/exhaustive-deps":             "off",
-		"react-hooks/rules-of-hooks":              "off",
+		"react-hooks/exhaustive-deps":             "error",
```
[the remark this answers is not in the document]
Comment (frosso) [reply .eslintrc:75]: I think it's best to keep it to `error`. `warn` can be easily overlooked, whereas with `error`, the `eslint` check would fail and it makes the developer more aware of the possible issue (which needs to be explicitly be disabled with a `eslint-disable` comment).
In my personal experience I always set this rule to `error`, rather than `warn` for this reason :)
And it also helped by making people look at the documentation 😆

File: client/settings/payment-methods-selector/index.js [hunk 15]
```diff
@@ -16,12 +16,12 @@ import PaymentMethodCheckboxes from '../../components/payment-methods-checkboxes
 import PaymentMethodCheckbox from '../../components/payment-methods-checkboxes/payment-method-checkbox';
 import './style.scss';
 
+const availablePaymentMethods = [
+	'woocommerce_payments_giropay',
+	'woocommerce_payments_sofort',
+	'woocommerce_payments_sepa',
+];
 const PaymentMethodsSelector = ( { className } ) => {
```
Comment (asumaran) [inline client/settings/payment-methods-selector/index.js:24]: is this change related to this PR?

File: client/utils/use-confirm-navigation.js (hunk 9 above)
Comment (asumaran) [reply client/utils/use-confirm-navigation.js:19]: I guess the "Usage" above is not correct anymore?

```
 * - useConfirmNavigation( saved => { if ( ! saved ) return 'Discard unsaved changes?' }, [ saved ] );
```

**Review by asumaran — CHANGES_REQUESTED** [review]
Overall looks good to me. I left a few comments for you @frosso.

Also, please note there still one lint error:

```
~/www/woocommerce-payments ‹frosso/enable-eslint-hooks-rules› » npm run lint:js                                                                                                                                                               1 ↵

> woocommerce-payments@2.4.0 lint:js /Users/asumaran/www/woocommerce-payments
> tsc --noEmit && eslint . --ext=js,jsx,ts,tsx


/Users/asumaran/www/woocommerce-payments/client/additional-methods-setup/wizard/wrapper/index.js
  25:3  error  React Hook useMemo has a duplicate dependency: 'completedTasks'. Either omit it or remove the dependency array  react-hooks/exhaustive-deps
```

File: client/payment-details/timeline/test/index.js [hunk 16]
```diff
@@ -170,14 +170,4 @@ describe( 'PaymentDetailsTimeline', () => {
 
 		expect( container ).toMatchSnapshot();
 	} );
-
-	test( 'does not render when the feature flag is disabled', () => {
```
[the remark this answers is not in the document]
Comment (frosso) [reply client/payment-details/timeline/test/index.js:174]: Yeah, I noticed the same thing - for that reason I didn't move it 😬 it looks like I would have had to write a bunch of stuff for it 😄

File: client/settings/payment-methods-selector/index.js (hunk 15 above)
Comment (frosso) [reply client/settings/payment-methods-selector/index.js:24]: Yes - the `availablePaymentMethods` is defined within the body of `PaymentMethodsSelector`.
Which means that its reference changes every time `PaymentMethodsSelector` runs. But `availablePaymentMethods` is used inside a hook. And since its reference changes at each render, and since it needs to be declared as a hook dependency, the hook would not be memoized properly.
By extracting it to a `const` outside of the body of `PaymentMethodsSelector`, its reference doesn't change.

File: client/utils/use-confirm-navigation.js (hunk 9 above)
Comment (frosso) [reply client/utils/use-confirm-navigation.js:19]: Ah, good catch, thanks! Fixed in 67db5624

**Review by frosso — COMMENTED** [review]
@asumaran thanks! Gotta love keep updating the PR as people use the rule incorrectly :D updated in 96b7719b1709847b80de1b0232fb0148705069f4

**Review by asumaran — APPROVED** [review]
LGTM
